### PR TITLE
Fixes for cookie signing

### LINKED ISSUE #154: Signing cookies problem
With configuration:
```
nelmio_security:
    # signs/verifies all cookies
    signed_cookie:
        names: ['symfony', 'two_factor_trusted_computer']
```
I get my system cannot login anymore for any user. Is there a requirement for this?

symfony configuration:
```
framework:
    #esi:             ~
    secret:          "%secret%"
    router:
        resource: "%kernel.root_dir%/config/routing.yml"
        strict_requirements: ~
    form: ~
    csrf_protection: ~
    validation:      { enable_annotations: false }
    translator:
        enabled: false
    templating:
        engines: ['twig', 'php']
    assets: false
    default_locale:  "%locale%"
    trusted_hosts:   ~
    session:
        name: "symfony"
        handler_id:  snc_redis.session.handler
    fragments:       ~
    http_method_override: true
    ide: "phpstorm://open?file=%%f&line=%%l"
```

### LINKED ISSUE #312: Problem with latest Symfony update and signed_cookie feature
After the latest Symfony update with this pull [https://github.com/symfony/symfony/pull/46249](https://github.com/symfony/symfony/pull/46249) Symfony refreshes the session on each page view because the implemented regex filter from the pull reequest fails because of the "." stored in the session cookie with signed_cookie feature enabled.

It would be easier to fix this one on Symfony side but maybe they keep the restricted check and an update in this library is needed...

### LINKED ISSUE #313: Incompatible with Symfony's clearCookie() strategy
Symfony's default method of deleting cookies is calling $response->headers->clearCookie('cookie_name') which sets the cookie to have a value of null and an expiration in 1 second.

This breaks if the cookie is signed, on line 79 of SignedCookieListener.php, when the code attempts to sign the "null" value.

`$this->signer->getSignedValue($cookie->getValue())` fails when $cookie->getValue() returns null.

I'm not entirely sure what the cleanest solution is. Either changing the above line to not try to sign them if the value is null (probably?). Or, maybe converting null values to a string?

For now, my workaround is to avoid clearCookie and to set the cookie to an expiration of 1 and a value of '', but this is messier than using the built-in Symfony method.

### PR DESCRIPTION
**Pull request by GwendolenLynch**
I hit a problem with CSRF tokens being rejected as invalid on both login and Symfony Form submissions. 

After a lot of head scratching, I observed that multiple sessions were being created per request and that for some reason the session ID was being invalidated by NativeSessionStorage. 

See https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php#L169-L172 and the comment above does the job explaining why.

The changes in Symfony were introduced here https://github.com/symfony/symfony/pull/46249 and recently updated here https://github.com/symfony/symfony/pull/46790

**tl;dr** Using a dot as the value delimiter when signing cookies, and signing session cookies, will cause sessions to fail.

Finally, if signing cookies with an empty value causes an exception. So the second fix just  handles that case.

Fixes #154 
Fixes #312 
Fixes #313

### CODE REVIEW

**Comment by yellow1912** [comment]
I'm running into this exact issue. Hope the PR is merged soon.

File: src/EventListener/SignedCookieListener.php [hunk 1]
```diff
@@ -72,7 +72,7 @@ public function onKernelResponse(ResponseEvent $e): void
         $response = $e->getResponse();
 
         foreach ($response->headers->getCookies() as $cookie) {
-            if (true === $this->signedCookieNames || \in_array($cookie->getName(), $this->signedCookieNames, true)) {
+            if ($cookie->getValue() && (true === $this->signedCookieNames || \in_array($cookie->getName(), $this->signedCookieNames, true))) {
```
Comment (Seldaek) [inline src/EventListener/SignedCookieListener.php:75]:
```suggestion
            if (null !== $cookie->getValue() && (true === $this->signedCookieNames || \in_array($cookie->getName(), $this->signedCookieNames, true))) {
```

File: src/Signer.php [hunk 2]
```diff
@@ -75,7 +75,7 @@ private function generateSignature(string $value): string
      */
     private function splitSignatureFromSignedValue(string $signedValue): array
     {
-        $pos = strrpos($signedValue, '.');
+        $pos = strrpos($signedValue, ',');
```
Comment (Seldaek) [inline src/Signer.php:78]: I'm not so sure about this.. On one hand it seems ok as a workaround, although I would like to still have BC here for `.` separated cookies so old cookies still work after deploying the new code (i.e. it should split using the last dot or comma found in the string if any).

On the other hand, it seems like the root cause is that you are signing session cookies, and https://github.com/nelmio/NelmioSecurityBundle/blob/7cc3a47ee6f57202ab53c5b1d0378e1a4e10d1a4/src/EventListener/SignedCookieListener.php#L57-L61 does not update $_COOKIE directly so the NativeSessionHandler does not see the unsigned-cookie value, and it actually uses the signed value as session id which to me sounds still borked.

So IMO we should probably write to $_COOKIE to unsign the cookies as well as writing to the request.. Any opinion here @nicolas-grekas from the Symfony/session perspective?

Comment (Seldaek) [reply src/Signer.php:78]: Alternative fix would be to NOT sign the session cookie ever even if `*` is configured for signed cookies, because really that id should anyway be verified in the session storage to prevent session fixation, so I don't think signing this actually improves anything security wise.

**Comment by WebDaMa** [comment]
@Seldaek What is the status of this PR? Our company is having the same bug.
If you can provide a todo list, we can work on it.

**Comment by gqai** [comment]
@WebDaMa also waiting for the PR, specifically the not signing a "null" cookie. We did find a work around in the interim, if you'd like to reach out.

@Seldaek is the only thing currently blocking this the two checks that are failing? Maybe I can take a look at it.

**Comment by Seldaek** [comment]
Nope it's not only the build, it's also my comments above which were not addressed.

**Comment by Seldaek** [comment]
https://github.com/nelmio/NelmioSecurityBundle/releases/tag/v3.2.0 has added support for null cookies - but the issue with signing the session cookie might still be current. I am not sure and I'd have to try it to confirm I guess.. But as time is very limited here, can someone tell me if this PR is still needed?

**Comment by OdendaalG** [comment]
@Seldaek If v3.2.0 resolves the null cookie issue, then it fixes the issues I came across. On the matter of the signing (which I never found to be a bug in the end), I believe that is unresolved but would require a bit of time to solve. Given the lack of time on your part and I think the prevalence, it could go to the back burner. Let me know if I can assist with something but in the mean time, thank you for this.
